### ISSUE TITLE
Unable to reproduce model results

### ISSUE THREAD

**Issue by jacob-zietek** [google-research/language-table#11]
Hello! Thank you for releasing the training script as requested in #5.

Unfortunately, I haven't had any luck reproducing the results from the paper using the provided training script and config. I slightly modified the config in `language_table/train/configs/language_table_sim_local.py` so that checkpoints save every 50,000 steps and the training stats print every 5,000 steps, otherwise it's all the same.

The loss decreases over the first 1,000 steps and then fluctuates around loss=0.9 and loss_std=0.5 for the remaining ~700,000 steps that I trained for. As expected, the model doesn't perform well using the evaluation script. 

Something strange that I noticed is that the LR stayed the same throughout the training, which is unexpected since the training uses the adam optimizer. 

I noticed that the batch size was commented out for TPUs, was this config built only to be used with TPUs?

Command: `nohup python language_table/train/main.py --config=./language_table/train/configs/language_table_sim_local.py --workdir={PATH}/data/language_table_train/ &`

Start of training:
![image](https://user-images.githubusercontent.com/33991406/227697415-e3b470cf-70bc-4b2c-8075-25ea50a949b7.png)

Start of output log:
![image](https://user-images.githubusercontent.com/33991406/227697627-a0c3dbcc-8ac3-42ce-b208-71d9c24a5e33.png)

I am using an NVIDIA GeForce RTX 3090, and I'm currently training again with a larger batch size.

Please let me know if you would like any additional info or the full training log. Thank you!

**Comment by jacob-zietek** [comment]
Edit: The loss seems to get stuck and the LR doesn't change when the batch size is 128 as well.

**Comment by yangjiangeyjg** [comment]
Hey Jacob, I have encountered exactly the same problem as you. My loss also quickly dropped to around 0.9 and hovered. My global batchsize is set to 64, and lowering the learning rate does not change this situation. Can you provide us with some help? Thank you very much! @ayzaan

**Comment by ayzaan** [comment]
Hello,
The actual models we train use a much larger batch size of 4096. The config file `language_table_sim_local.py` is mostly intended as a reference for testing that the training runs locally.

A config that matches performance from the paper has the following:
```
config.data_normalization_num_samples = 3200
# 128 chips; global batch size = 4096; 32 per device.
config.per_device_batch_size = 32
config.replay_capacity = 50_000
```

The per_device_batch_size is 32 with 128 chips total (8x8 TPU), giving global batch size of 4096. In the experiments we ran for the paper (Table VI), cutting the batch size in half dropped performance by about 5%, so wouldn't be surprised if dropping it to 128 had a much larger effect. @jacob-zietek Do you have an idea of the largest batch size that you could fit on a 3090? If you'd like I can try to launch a training run with that batch size as a reference point for you to reproduce. Best would be to use 4096 but that will probably require more accelerator resources.

Also, wouldn't hurt to increase both the data_normalization_num_samples and replay_capacity, those values above are what we used in the paper. I'll check in a template config with these values for reference.

Let me know if you have any more questions.

**Comment by jacob-zietek** [comment]
Hi @ayzaan, thanks for the quick response!

Thank you for clearing that up. Unfortunately, the largest batch size that can fit on a single 3090 is 128. it peaks at ~22.5GB of VRAM during training. I would appreciate it if you could try to train with 128!

I see the replay_capacity is tied to the shuffle buffer size [here](https://github.com/google-research/language-table/blob/a4ffffdd623d6bc9fe1fca7767ae5a7be9a4b5cf/language_table/train/input_pipeline_rlds.py#L159). Shouldn't this be set to (Devices * Batch size) instead of 50_000? Perhaps I'm misunderstanding the purpose of the buffer.

I will try to train with a larger data_normalization_num_samples and replay_capacity now, thank you again!

**Comment by ayzaan** [comment]
`replay_capacity` is perhaps not the best name, it does just mean the shuffle buffer size in the data pipeline - to ensure the input data is shuffled adequately. We also trained in a distributed setting and had distributed data workers, which each shuffled with 50k elements and fed into training. In a smaller setup like yours I'd probably try to crank it as high as possible while still fitting on your machine RAM.

I'll launch a training run with a small batch size like 128 and see what happens. One hunch is that you should also decrease the learning rate significantly.

**Comment by jacob-zietek** [comment]
I see, thanks for clearing that up! I'll experiment with increasing the capacity and decreasing the LR if this training run fails.

**Comment by yangjiangeyjg** [comment]
> `replay_capacity` is perhaps not the best name, it does just mean the shuffle buffer size in the data pipeline - to ensure the input data is shuffled adequately. We also trained in a distributed setting and had distributed data workers, which each shuffled with 50k elements and fed into training. In a smaller setup like yours I'd probably try to crank it as high as possible while still fitting on your machine RAM.
> 
> I'll launch a training run with a small batch size like 128 and see what happens. One hunch is that you should also decrease the learning rate significantly.
Thank you for detailed explanation. Can you share the approximate convergence value of loss when you train on language_table and language_table_sim. Thanks!

**Comment by yangjiangeyjg** [comment]
Thank you for detailed explanation. Can you share the approximate convergence value of loss when you train on language_table and language_table_sim. Thanks! @ayzaan

**Comment by ayzaan** [comment]
Here are a couple of results on language_table_sim. Overall it seems performance does decrease significantly with smaller batches, although you should still see some successful rollouts during eval. 

black is the baseline with batch_size=4096
orange is batch_size=256 with LR=1e-4
green is batch_size=128 with LR=5e-5.
<img width="399" alt="Screenshot 2023-03-27 at 5 50 47 PM" src="https://user-images.githubusercontent.com/1704564/228098866-f6e81092-baa2-421d-84a9-112eee7e957a.png">

And success rates:

<img width="390" alt="Screenshot 2023-03-27 at 5 50 31 PM" src="https://user-images.githubusercontent.com/1704564/228099356-4e9709f6-038f-4b5d-a2aa-a0e6340f9694.png">

**Comment by jacob-zietek** [comment]
Thanks a lot for running additional training runs! I will use batch_size=128 with LR=5e-5
